sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/OptimizeSkewedPartitions.scala

```diff
@@ -0,0 +1,253 @@
+      val numPartitions = leftStats.bytesByPartitionId.length
+
+      val leftMedSize = medianSize(leftStats)
+      val rightMedSize = medianSize(rightStats)
+      logDebug(
+        s"""
+          |Try to optimize skewed join.
+          |Left side partition size: median size: $leftMedSize,
+          | max size: ${leftStats.bytesByPartitionId.max}
+          |Right side partition size: median size: $rightMedSize,
+          | max size: ${rightStats.bytesByPartitionId.max}
+        """.stripMargin)
+
+      val skewedPartitions = mutable.HashSet[Int]()
+      val subJoins = mutable.ArrayBuffer[SparkPlan]()
+      for (partitionId <- 0 until numPartitions) {
+        val isLeftSkew = isSkewed(leftStats, partitionId, leftMedSize)
+        val isRightSkew = isSkewed(rightStats, partitionId, rightMedSize)
+        val leftMapIdStartIndices = if (isLeftSkew && supportSplitOnLeftPartition(joinType)) {
+          getMapStartIndices(left, partitionId, leftMedSize)
+        } else {
+          Array(0)
+        }
+        val rightMapIdStartIndices = if (isRightSkew && supportSplitOnRightPartition(joinType)) {
+          getMapStartIndices(right, partitionId, rightMedSize)
+        } else {
+          Array(0)
+        }
+
+        if (leftMapIdStartIndices.length > 1 || rightMapIdStartIndices.length > 1) {
+          skewedPartitions += partitionId
+          for (i <- 0 until leftMapIdStartIndices.length;
+               j <- 0 until rightMapIdStartIndices.length) {
+            val leftEndMapId = if (i == leftMapIdStartIndices.length - 1) {
+              getNumMappers(left)
+            } else {
+              leftMapIdStartIndices(i + 1)
+            }
+            val rightEndMapId = if (j == rightMapIdStartIndices.length - 1) {
+              getNumMappers(right)
+            } else {
+              rightMapIdStartIndices(j + 1)
+            }
+            // TODO: we may can optimize the sort merge join to broad cast join after
+            //       obtaining the raw data size of per partition,
+            val leftSkewedReader =
+            SkewedShufflePartitionReader(
+              left, partitionId, leftMapIdStartIndices(i), leftEndMapId)
+
+            val rightSkewedReader =
+              SkewedShufflePartitionReader(right, partitionId,
+                rightMapIdStartIndices(j), rightEndMapId)
+            subJoins += SortMergeJoinExec(leftKeys, rightKeys, joinType, condition,
+              leftSkewedReader, rightSkewedReader)
+          }
+        }
+      }
+      logDebug(s"number of skewed partitions is ${skewedPartitions.size}")
+      if (skewedPartitions.size > 0) {
+        val optimizedSmj = smj.transformDown {
+          case sort: SortExec if (
+            ShuffleQueryStageExec.isShuffleQueryStageExec(sort.child)) => {
+            val newStage = ShuffleQueryStageExec.getShuffleStage(
+              sort.child.asInstanceOf[QueryStageExec]).copy(
+              excludedPartitions = skewedPartitions.toSet)
```

Comment:
This always return a `ShuffleQueryStageExec`, but the original plan may be a `ReusedQueryStageExec`. We shouldn't strip the reuse plan.

we should do something like
```
sort.child match {
  case s: ShuffleQueryStageExec => s.copy...
  case r @ ReusedQueryStageExec(s: ShuffleQueryStageExec) => r.copy(s.copy...)
}
```